event/buffered_bus.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package event
+
+import (
+	"sync/atomic"
+)
+
+type subscribeCommand *Subscription
+
+type unsubscribeCommand *Subscription
+
+type publishCommand Message
+
+type closeCommand struct{}
+
+// bufferedBus must implement the Bus interface
+var _ (Bus) = (*bufferedBus)(nil)
+
+// bufferedBus is a bus that uses a buffered channel to manage subscribers and publish messages.
+type bufferedBus struct {
+	// subID is incremented for each subscriber and used to set subscriber ids.
+	subID atomic.Uint64
+	// subs is a mapping of subscriber ids to subscriptions.
+	subs map[uint64]*Subscription
+	// events is a mapping of event names to subscriber ids.
+	events map[Name]map[uint64]struct{}
+	// commandChannel manages all commands sent to the bufferedBus.
+	//
+	// It is important that all stuff gets sent through this single channel to ensure
+	// that the order of operations is preserved.
+	//
+	// This does mean that non-event commands can block the database if the buffer
+	// size is breached (e.g. if many subscribe commands occupy the buffer).
+	commandChannel  chan any
+	eventBufferSize int
+	hasClosedChan   chan struct{}
+	isClosed        atomic.Bool
+}
+
+// NewBufferedBus creates a new event bus with the given commandBufferSize and
+// eventBufferSize.
+//
+// Should the buffers be filled, subsequent calls on this bus will block.
+func NewBufferedBus(commandBufferSize int, eventBufferSize int) *bufferedBus {
+	bus := bufferedBus{
+		subs:            make(map[uint64]*Subscription),
+		events:          make(map[Name]map[uint64]struct{}),
+		commandChannel:  make(chan any, commandBufferSize),
+		hasClosedChan:   make(chan struct{}),
+		eventBufferSize: eventBufferSize,
+	}
+	go bus.handleChannel()
+	return &bus
+}
+
+func (b *bufferedBus) Publish(msg Message) {
+	if b.isClosed.Load() {
+		return
+	}
+	b.commandChannel <- publishCommand(msg)
+}
+
+func (b *bufferedBus) Subscribe(events ...Name) (*Subscription, error) {
+	if b.isClosed.Load() {
+		return nil, ErrSubscribedToClosedChan
+	}
+	sub := &Subscription{
+		id:     b.subID.Add(1),
+		value:  make(chan Message, b.eventBufferSize),
+		events: events,
+	}
+	b.commandChannel <- subscribeCommand(sub)
+	return sub, nil
+}
+
+func (b *bufferedBus) Unsubscribe(sub *Subscription) {
+	if b.isClosed.Load() {
+		return
+	}
+	b.commandChannel <- unsubscribeCommand(sub)
+}
+
+func (b *bufferedBus) Close() {
+	if !b.isClosed.CompareAndSwap(false, true) {
+		return
+	}
+	b.commandChannel <- closeCommand{}
+	// Wait for the close command to be handled, in order, before returning
+	<-b.hasClosedChan
```

Comment:
question: this is thread-safe only under the assumption that all the calls to `bufferedBus` are done by the same thread. I'm not sure this is the case. Is it?

Otherwise the following scenario is possible:
1. execution enters `Subscribe` method, `b.isClosed == false` so it proceeds but pauses just before `b.commandChannel <- subscribeCommand(sub)`
2. concurrently `Close` is being called, sets `b.isClosed` to `true`, pushes `b.commandChannel <- closeCommand{}`
3. `handleChannel` receives `closeCommand` and executes `close(b.commandChannel)`
4. now the thread from step 1 resumes, executes `b.commandChannel <- subscribeCommand(sub)` and causes a panic.

Reply:
That's quite a broad mutex, please don't merge yet

Reply:
It should only block when calling close.

Reply:
Okay, I think I understand, and there is no nicer way of handling this that I can spot off the top of my head.

todo: Can you please rename `mutex` to `closeLock` or similar, we really really don't want this to be used for anything else.

todo: Can please also document the lock on its declaration in `bufferedBus`, explaining that it is only locked whilst closing.

Reply:
@AndrewSisley can you explain what you mean by broad? We use the same technique in the memory store and I think it's really fine for this as well.

Reply:
> @AndrewSisley can you explain what you mean by broad? We use the same technique in the memory store and I think it's really fine for this as well.

The lock spans every public call to this struct, it can't really get any broader :)  If it only locks whilst closing that's fine, but it would be very damaging if that were to change (hence the documentation and name change requested).